sdk/src/main/scala/com/horizen/forge/ForgeMessageBuilder.scala

```diff
@@ -55,47 +62,129 @@ class ForgeMessageBuilder(mainchainSynchronizer: MainchainSynchronizer,
+    val parentBlockId: ModifierId = branchPointInfo.branchPointId
+    val parentBlockInfo: SidechainBlockInfo = view.history.blockInfoById(branchPointInfo.branchPointId)
+    var withdrawalEpochMcBlocksLeft: Int = params.withdrawalEpochLength - parentBlockInfo.withdrawalEpochInfo.lastEpochIndex
+    if (withdrawalEpochMcBlocksLeft == 0) // parent block is the last block of the epoch
       withdrawalEpochMcBlocksLeft = params.withdrawalEpochLength
 
-    val mainchainBlockRefToInclude: Seq[MainchainBlockReference] = mainchainSynchronizer.getNewMainchainBlockReferences(
-      view.history,
-      Math.min(SidechainBlock.MAX_MC_BLOCKS_NUMBER, withdrawalEpochMcBlocksLeft) // to not to include mcblock references from different withdrawal epochs
-    )
+    // Get all needed MainchainBlockReferences from MC Node
+    val allMainchainHashes: Seq[MainchainHeaderHash] = (branchPointInfo.referenceDataToInclude ++ branchPointInfo.headersToInclude).distinct
+    val mainchainBlockReferences: Seq[MainchainBlockReference] =
+      mainchainSynchronizer.getMainchainBlockReferences(view.history, allMainchainHashes) match {
+        case Success(references) => references
+        case Failure(ex) => return ForgeFailed(ex)
+      }
+
+    // Extract proper MainchainReferenceData
+    val mainchainReferenceData: Seq[MainchainBlockReferenceData] =
+      mainchainBlockReferences.filter(ref => branchPointInfo.referenceDataToInclude.contains(byteArrayToMainchainHeaderHash(ref.header.hash)))
+      .map(_.data)
 
-    val txsToInclude: Seq[SidechainTransaction[Proposition, NoncedBox[Proposition]]] =
-      if(mainchainBlockRefToInclude.size == withdrawalEpochMcBlocksLeft) { // SC block is going to become the last block of the withdrawal epoch
+    // Extract proper MainchainHeaders
+    val mainchainHeaders: Seq[MainchainHeader] =
+      mainchainBlockReferences.filter(ref => branchPointInfo.headersToInclude.contains(byteArrayToMainchainHeaderHash(ref.header.hash)))
+        .map(_.header)
+
+    // Get transactions if possible
+    val transactions: Seq[SidechainTransaction[Proposition, NoncedBox[Proposition]]] =
+      if (branchPointInfo.referenceDataToInclude.size == withdrawalEpochMcBlocksLeft) { // SC block is going to become the last block of the withdrawal epoch
         Seq() // no SC Txs allowed
       } else { // SC block is in the middle of the epoch
         view.pool.take(SidechainBlock.MAX_SIDECHAIN_TXS_NUMBER) // TO DO: problems with types
           .map(t => t.asInstanceOf[SidechainTransaction[Proposition, NoncedBox[Proposition]]])
           .toSeq
       }
 
-    val blockCreationResult = SidechainBlock.create(
+    // Get ommers in case if branch point is not current best block
+    var ommers: Seq[Ommer] = Seq()
+    var blockId = view.history.bestBlockId
+    while (blockId != branchPointInfo.branchPointId) {
```

Comment:
we could creete custom scala stream from block to genesis blocks over blocks in history, in such case we can rephrase it into getBlocksSteam(view.history.bestBlockId).takeWhile(_ != branchPointInfo.branchPointId).map(Ommer.toOmmer)